File: Go/binaryInsertionSort.go
package main

import (
	"fmt"
)

// binarySearch is a helper function that finds the index to insert a given element in a sorted slice
// using binary search algorithm. It returns the index at which the element should be inserted.
func binarySearch(arr []int, item int, low int, high int) int {
	for low <= high {
		mid := low + (high-low)/2 // Find the mid-point to reduce the search space
		if arr[mid] < item {
			low = mid + 1 // Move to the right half if the mid-value is less than the item
		} else {
			high = mid - 1 // Move to the left half if the mid-value is greater than or equal to the item
		}
	}
	return low // Return the index where the item should be inserted
}

// binaryInsertionSort sorts an array using the binary insertion sort algorithm.
func binaryInsertionSort(arr []int) {
	// Loop through each element starting from the second element (index 1)
	for i := 1; i < len(arr); i++ {
		key := arr[i] // The current element to be positioned
		j := i - 1    // The last index of the sorted part of the array

		// Find the location where the key should be inserted using binary search
		insertPos := binarySearch(arr, key, 0, j)

		// Shift elements to the right to make space for the key
		for j >= insertPos {
			arr[j+1] = arr[j] // Shift the element at j to the right
			j--               // Move to the previous index
		}
		arr[insertPos] = key // Insert the key at its correct position
	}
}

// main function to test the binaryInsertionSort
func main() {
	arr := []int{12, 11, 13, 5, 6} // Sample array to be sorted
	fmt.Println("Original array:", arr)

	binaryInsertionSort(arr) // Call the sorting function

	fmt.Println("Sorted array:", arr) // Print the sorted array
}

File: Java/binaryInsertionSort.java
public class BinaryInsertionSort {
    
    // Method to perform binary insertion sort
    public static void binaryInsertionSort(int[] array) {
        // Traverse through the array starting from the second element
        for (int i = 1; i < array.length; i++) {
            // The current element to be positioned
            int key = array[i];
            // Find the position to insert the key
            int position = binarySearch(array, key, 0, i - 1);

            // Shift elements to the right to make space for the key
            for (int j = i - 1; j >= position; j--) {
                array[j + 1] = array[j];
            }
            // Place the key at its correct position
            array[position] = key;
        }
    }

    // Method to perform binary search to find the correct position of the key
    private static int binarySearch(int[] array, int key, int low, int high) {
        // If the low index is greater than high, return the low index
        if (low > high) {
            return low;
        }

        // Calculate the mid index
        int mid = (low + high) / 2;

        // If key is less than the mid element, search in the left half
        if (key < array[mid]) {
            return binarySearch(array, key, low, mid - 1);
        }
        // If key is greater than the mid element, search in the right half
        else {
            return binarySearch(array, key, mid + 1, high);
        }
    }

    // Method to print the array
    public static void printArray(int[] array) {
        // Iterate through the array and print each element
        for (int value : array) {
            System.out.print(value + " ");
        }
        System.out.println();
    }

    // Main method to test the sorting algorithm
    public static void main(String[] args) {
        // Sample array to be sorted
        int[] array = {12, 11, 13, 5, 6};

        System.out.println("Original array:");
        printArray(array); // Print the original array

        // Call the binary insertion sort method
        binaryInsertionSort(array);

        System.out.println("Sorted array:");
        printArray(array); // Print the sorted array
    }
}
